apps/learner/src/lib/helpers/is-within-viewport.svelte.ts

```diff
@@ -0,0 +1,46 @@
+/**
+ * Tracks if an element is within the current viewport.
+ *
+ * @example
+ * ```svelte
+ * <script lang="ts">
+ * let target = $state<HTMLElement | null>(null);
+ *
+ * const isWithinViewport = new IsWithinViewport(() => target);
+ * </script>
+ *
+ * <span bind:this={target}>Hello world!</span>
+ * <span>Is within viewport: {isWithinViewport.current ? 'Yes' : 'No'}</span>
+ * ```
+ */
+export class IsWithinViewport {
+  #isWithinViewport = $state(false);
+
+  constructor(target: () => HTMLElement | null) {
+    let observer: IntersectionObserver | null = null;
+
+    $effect(() => {
+      observer = new IntersectionObserver(([entry]) => {
+        this.#isWithinViewport = entry.isIntersecting;
+      });
+
+      const element = target();
```

Comment:
Do we need null checks or can we safely assume that target() always returns a valid element?

Reply:
We cannot assume that the target would not be null. It is possible that the target is null.

Reply:
The next line is a null check.

Reply:
Thought of adding sufficient checks to avoid stale state. Can proceed with this for now.